ui/src/common/actions.ts

```diff
@@ -143,6 +160,88 @@ function removeTrack(state: StateDraft, trackId: string) {
+  delete state.trackGroups[groupId];
+
+  state.pinnedTracks = state.pinnedTracks.filter((id) => id !== groupId);
+}
+
+function hasRemovable<T extends AddTrackLikeArgs>(item: T): item is T & {isRemovable: boolean} {
+  return 'isRemovable' in item && typeof item.isRemovable === 'boolean';
+}
+
+function wasFiltered<T extends TrackState|TrackGroupState>(item: T): item is T & {wasFiltered?: true} {
+  return 'wasFiltered' in item && item.wasFiltered === true;
+}
+
+function isFilteredTrack(track: Partial<AddTrackArgs>): boolean {
+  return globals.trackFilteringEnabled && globals.filteredTracks
+      .some(filtered => isAddTrackArgs(filtered) && isSameTrack(filtered, track));
+}
+
+// Query whether an |other| track matches enough details of a |track| as
+// to represent the same track
+function isSameTrack(track: AddTrackArgs, other: Partial<AddTrackArgs>): boolean {
+  return track.kind === other.kind
+      && track.trackGroup == other.trackGroup
+      // TODO: This may not be reliable. May need to deep-compare the config object
+      && track.name == other.name;
+}
+
+function isFilteredTrackGroup(trackGroup: Partial<AddTrackGroupArgs>): boolean {
+  return globals.trackFilteringEnabled && globals.filteredTracks
+    .some(filtered => isAddTrackGroupArgs(filtered) && filtered.id === trackGroup.id);
+}
+
+function unfilterTracklike(predicate: (tracklike: AddTrackLikeArgs) => boolean) {
+  const index = globals.filteredTracks.findIndex(predicate);
+  if (index >= 0) {
+    globals.filteredTracks.splice(index, 1);
+  }
+}
+
+function unfilterTrack(track: TrackState) {
+  track.isRemovable = true;
+  (track as any).wasFiltered = true;
+  unfilterTracklike(filtered => isAddTrackArgs(filtered) && isSameTrack(filtered, track));
+}
+
+function unfilterTrackGroup(trackGroup: TrackGroupState) {
+  trackGroup.isRemovable = true;
+  (trackGroup as any).wasFiltered = true;
+  unfilterTracklike(filtered => isAddTrackGroupArgs(filtered) && filtered.id === trackGroup.id);
+}
+
+// A helper to delete the private tables and views created by a track.
+// TODO: These should recorded by each track that creates them and cleaned up
+//       by an explicit disposable-track protocol.
+async function dropTables(engineId: string, trackId: string) {
+  const engine = assertExists(globals.engines.get(engineId));
+  const suffix = trackId.split('-').join('_');
+  const result = await engine.query(`
+      select name, type from sqlite_schema
+      where name like '%_${suffix}'
+      union select name, type from sqlite_temp_schema
+      where name like '%_${suffix}'`);
+
+  const it = result.iter({name: STR, type: STR});
+  const dropStmts: string[] = [];
+  for (; it.valid(); it.next()) {
+    dropStmts.push(`drop ${it.type} ${it.name};`);
+  }
+
+  for (const stmt of dropStmts) {
+    try {
+      await engine.query(stmt);
+    } catch (_error) {
+      // This is expected, depending on the order in which
```

Comment:
Is this expected so much that we do not even log it or should we at least log it as something?

Reply:
Yes, that's right. The order in which we get the results is more-or-less non-deterministic and so most of the time you will see one or more logs that are unhelpful and non-actionable. Dropping these tables cannot otherwise fail and, if it does, then it's because already nothing in the database is working.